src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -1133,13 +1149,13 @@ public SelectStatement prepare(ClientState state, boolean forView) throws Invali
             List<Selectable> selectables = RawSelector.toSelectables(selectClause, table);
             boolean containsOnlyStaticColumns = selectOnlyStaticColumns(table, selectables);
 
-            StatementRestrictions restrictions = prepareRestrictions(state, table, bindVariables, containsOnlyStaticColumns, forView);
+            List<Ordering> orderings = getOrderings(table);
+            StatementRestrictions restrictions = prepareRestrictions(state, table, bindVariables, orderings, containsOnlyStaticColumns, forView);
 
             // If we order post-query, the sorted column needs to be in the ResultSet for sorting,
             // even if we don't ultimately ship them to the client (CASSANDRA-4911).
-            Map<ColumnMetadata, Boolean> orderingColumns = getOrderingColumns(table);
-            Set<ColumnMetadata> resultSetOrderingColumns = restrictions.keyIsInRelation() ? orderingColumns.keySet()
-                                                                                          : Collections.emptySet();
+            Map<ColumnMetadata, Ordering> orderingColumns = getOrderingColumns(orderings);
+            Set<ColumnMetadata> resultSetOrderingColumns = getResultSetOrdering(restrictions, orderingColumns);
```

Comment:
It seems that somehow the vector column of an ANN query is always included in the results set, even if it's not included in the selection. This single-node dtest shows the issue:
```java
try (Cluster cluster = init(Cluster.build(1).start()))
{
 cluster.schemaChange(withKeyspace("CREATE TABLE %s.t (k int PRIMARY KEY, v vector<float, 1>, c int)"));
 cluster.schemaChange(withKeyspace("CREATE CUSTOM INDEX ON %s.t(v) USING 'StorageAttachedIndex'"));

 cluster.coordinator(1).execute(withKeyspace("INSERT INTO %s.t (k, v, c) VALUES (0, [1], 2)"), ConsistencyLevel.QUORUM);

 String query = withKeyspace("SELECT c FROM %s.t ORDER BY v ANN OF [1] LIMIT 10");
 assertRows(cluster.coordinator(1).execute(query, ConsistencyLevel.QUORUM), row(2)); // includes the vector column!!
}
```
We can also reproduce it with a unit test, but it requires to define a page size, as I think drivers usually do:
```java
@Test
public void testSelection() throws Throwable
{
    createTable("CREATE TABLE %s (k int PRIMARY KEY, v vector<float, 1>, c int)");
    createIndex("CREATE CUSTOM INDEX ON %s(v) USING 'StorageAttachedIndex'");
    execute("INSERT INTO %s (k, v, c) VALUES (0, [1], 2)");

    String query = "SELECT c FROM %s ORDER BY v ANN OF [1] LIMIT 10";
    assertRows(execute(query), row(2));
    
    requireNetwork();
    assertRowsNet(executeNet(query), row(2));
    assertRowsNet(executeNetWithPaging(query, 1000), row(2)); // wrongly includes the vector column
}
```

Reply:
This failure is reproducible without SAI involved at all. The following test also fails on the `cassandra-5.0` branch:
```
@Test
public void test2() throws IOException
{
    try (Cluster cluster = init(Cluster.build(1).start()))
    {
        cluster.schemaChange(withKeyspace("CREATE TABLE %s.t (k int, c int, v int, primary key(k, c))"));

        cluster.coordinator(1).execute(withKeyspace("INSERT INTO %s.t (k, c, v) VALUES (0, 1, 2)"), ConsistencyLevel.QUORUM);

        String query = withKeyspace("SELECT v FROM %s.t WHERE k IN (0, 1) ORDER BY c LIMIT 10");
        assertRows(cluster.coordinator(1).execute(query, ConsistencyLevel.ONE), row(2));
    }
}
```

Reply:
I think both of these make sense because without the ordering column being returned it would be impossible to determine the ordering logic. If the above query is changed to `SELECT v FROM %s.t WHERE k = 0 ORDER BY c LIMIT 10`, then the clustering key is not returned.

Reply:
This is a bug in the distributed test code. Specifically, `RowUtil.toQueryResult`. This is incorrectly using `rows.result.metadata.names` instead of `rows.result.metadata.requestNames()`. The `requestNames` method specifically removes any columns that have not been requested. I have raised CASSANDRA-18892 and attached a fix patch there for this.

Reply:
Good catch, I have +1ed CASSANDRA-18892. I'll try to commit it tomorrow.